vowpalwabbit/core/src/reductions/eigen_memory_tree.cc

```diff
@@ -0,0 +1,928 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "vw/core/reductions/eigen_memory_tree.h"
+
+#include "vw/common/future_compat.h"
+#include "vw/config/options.h"
+#include "vw/core/array_parameters.h"
+#include "vw/core/example.h"
+#include "vw/core/learner.h"
+#include "vw/core/numeric_casts.h"
+#include "vw/core/rand48.h"
+#include "vw/core/rand_state.h"
+#include "vw/core/setup_base.h"
+#include "vw/core/v_array.h"
+#include "vw/core/vw.h"
+#include "vw/io/logger.h"
+
+#include <algorithm>
+#include <cfloat>
+#include <climits>
+#include <cmath>
+#include <cstdio>
+#include <ctime>
+#include <list>
+#include <memory>
+#include <sstream>
+
+using namespace VW::LEARNER;
+using namespace VW::config;
+
+// TODO: This file has several cout print statements. It looks like
+//       they should be using trace_message, but its difficult to tell
+namespace VW
+{
+namespace reductions
+{
+namespace eigen_memory_tree
+{
+///////////////////////Helper/////////////////////////////
+//////////////////////////////////////////////////////////
+float median(std::vector<float>& array)
+{
+  int size = array.size();
+  std::sort(array.begin(), array.end());
```

Comment:
is it beneficial to have linear complexity algorithm here?
i.e. stl ones: https://cplusplus.com/reference/algorithm/nth_element/

Reply:
Oh, nice, I didn't know about that method. That's an easy change to make to be safe. Jack had also mentioned maybe there was an existing algorithm in VW that is even better? For what it is worth, in practice I don't think it'll matter too much. We only calculate the median when we split a node in our binary tree, which means it's calculated with log(N) frequency. And when we do calculate it is only on the items in the leaf node being split.